### PR TITLE
Feat: Move Sharing State Transition Logic to the Sharing State Service

### PR DESCRIPTION
**Pull request by nicoprow**
## Description

Currently, the sharing state service is somewhat of a glorified repository. It just takes a DTO, maps it to the entity and saves whatever is inside. This means the logic of managing sharing state transitions is left to the invoking services.

The goal of this pull request is to move such sharing state transition logic to the sharing state, by this capsulating responsibitilies between the services better.

In addition I added some validation logic for transitioning to shring states making sure essential information for that state is given. At the same time, information that is irrelevant for the transitioned sharing state is being ignored (such as error information for any other state than Error).

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

File: bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt
Comment (martinfkaeser) [inline bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt:82]: Maybe use `.toPageDto { SharingStateDto(...) }` from `CommonMappings.kt`?

File: bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt
Comment (martinfkaeser) [inline bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt:102]: Variable name is confusing.
```suggestion
        val bpnByExternalId = externalIdBpnDto.associateBy { it.externalId }.mapValues { it.value.bpn }
```

File: bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt
Comment (nicoprow) [reply bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt]: I added it via a simple logic to map from business partner types to the relevant BPN.

Later on, we should move from setting BPN and business partner type explicitely in the sharing state and rather inferring it from the business partner stages. But this can only happen after we integrated the L/S/A endpoints to the generic model, made the external identifiers unique and the business partner type optional in the sharing state.

File: bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt
Comment (nicoprow) [reply bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/SharingStateService.kt]: Yep, makes sense

File: bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt
Comment (nicoprow) [reply bpdm-gate/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/service/BusinessPartnerService.kt]: I added it via a simple logic to map from business partner types to the relevant BPN.

Later on, we should move from setting BPN and business partner type explicitely in the sharing state and rather inferring it from the business partner stages. But this can only happen after we integrated the L/S/A endpoints to the generic model, made the external identifiers unique and the business partner type optional in the sharing state.

**Review by martinfkaeser — APPROVED** [review]

**Review by fabiodmota — APPROVED** [review]

**Comment by nicoprow** [comment]
Wait until #519 is merged before this should be merged

**Pull request merged**
